components/tor/tor_launcher_factory.cc

```diff
@@ -260,40 +274,61 @@ void TorLauncherFactory::GotSOCKSListeners(
     observer.OnTorNewProxyURI(tor_proxy_uri);
 }
 
-void TorLauncherFactory::OnTorClosed() {
-  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+void TorLauncherFactory::OnTorControlClosed(bool was_running) {
+  SEQUENCE_CHECKER(sequence_checker_);
   VLOG(2) << "TOR CONTROL: Closed!";
+  // If we're still running, try watching again to start over.
+  // XXX Rate limit in case of flapping?
+  if (was_running) {
+    LaunchTorInternal();
+  }
 }
 
-void TorLauncherFactory::OnTorCleanupNeeded(base::ProcessId id) {
-  DCHECK_CURRENTLY_ON(BrowserThread::UI);
-  VLOG(2) << "Killing old tor process pid=" << id;
-  // Dispatch to launcher thread
-  content::GetProcessLauncherTaskRunner()->PostTask(
-      FROM_HERE, base::BindOnce(&TorLauncherFactory::KillOldTorProcess,
-                                base::Unretained(this), std::move(id)));
-}
-
-void TorLauncherFactory::KillOldTorProcess(base::ProcessId id) {
-  DCHECK(content::CurrentlyOnProcessLauncherTaskRunner());
-  base::Process tor_process = base::Process::Open(id);
-  if (tor_process.IsValid())
-    tor_process.Terminate(0, false);
+void TorLauncherFactory::OnTorControlPrerequisitesReady(
+    int64_t pid,
+    bool ready,
+    std::vector<uint8_t> cookie,
+    int port) {
+  SEQUENCE_CHECKER(sequence_checker_);
+  if (pid != tor_pid_) {
+    VLOG(1) << "Tor control pid mismatched!";
+    return;
+  }
+  if (ready) {
+    control_->Start(std::move(cookie), port);
+  } else {
+    tor_file_watcher_ = new tor::TorFileWatcher(config_.tor_watch_path);
+    tor_file_watcher_->StartWatching(base::BindPostTask(
+        base::SequencedTaskRunnerHandle::Get(),
+        base::BindOnce(&TorLauncherFactory::OnTorControlPrerequisitesReady,
+                       weak_ptr_factory_.GetWeakPtr(), pid)));
+  }
 }
 
 void TorLauncherFactory::RelaunchTor() {
+  SEQUENCE_CHECKER(sequence_checker_);
   Init();
-  control_->PreStartCheck(
-      config_.tor_watch_path,
-      base::BindOnce(&TorLauncherFactory::OnTorControlCheckComplete,
-                     weak_ptr_factory_.GetWeakPtr()));
+  LaunchTorInternal();
+}
+
+void TorLauncherFactory::DelayedRelaunchTor() {
+  SEQUENCE_CHECKER(sequence_checker_);
+  is_starting_ = false;
+  is_connected_ = false;
+  KillTorProcess();
+  // Post delayed relaunch for control to stop
+  content::GetUIThreadTaskRunner({})->PostDelayedTask(
```

Comment:
I'd rather use `base::SequencedTaskRunnerHandle::Get()` for consistency

Reply:
fixed in 47093e7b5b5e8e93172696c68a52086fb3a1aaf6